aspnetcore/performance/blazor/webassembly-best-practices.md

```diff
@@ -0,0 +1,156 @@
+
+If authoring a UI-only component that never changes after the initial render, configure `ShouldRender` to return `false`:
+
+```razor
+@code {
+    protected override bool ShouldRender() => false;
+}
+```
+
+Also use <xref:Microsoft.AspNetCore.Components.ComponentBase.ShouldRender%2A> to selectively render a component responding to a UI event.
+
+In the following example:
+
+* <xref:Microsoft.AspNetCore.Components.ComponentBase.ShouldRender%2A> is overridden and set to the value of the `shouldRender` field, which is initially `false` when the component loads.
+* When the button is selected, `shouldRender` is set to `true`, which forces the component to rerender with the updated `currentCount`.
+* Immediately after rerendering, <xref:Microsoft.AspNetCore.Components.ComponentBase.OnAfterRender%2A> sets the value of `shouldRender` back to `false` to prevent further rerendering until the next time the button is selected.
+
+```razor
+<p>Current count: @currentCount</p>
+
+<button @onclick="IncrementCount">Click me</button>
+
+@code {
+    private int currentCount = 0;
+    private bool shouldRender;
+
+    protected override bool ShouldRender() => shouldRender;
+
+    protected override void OnAfterRender(bool first)
+    {
+        shouldRender = false;
+    }
+
+    private void IncrementCount()
+    {
+        currentCount++;
+        shouldRender = true;
+    }
+}
+```
+
+For more information, see <xref:blazor/lifecycle#after-component-render>.
+
+## Virtualize re-usable fragments
+
+Components offer a convenient approach to produce re-usable fragments of code and markup. In general, we recommend authoring individual components that best align with the app's requirements. One caveat is that each additional child component contributes to the total time it takes to render a parent component. For most apps, the additional overhead is negligible. Apps that produce a large number of components should consider using strategies to reduce processing overhead, such as limiting the number of rendered components.
+
+For example, a grid or list that renders hundreds of rows containing components is processor intensive to render. Consider virtualizing a grid or list layout so that only a subset of the components is rendered at any given time. For an example of component subset rendering, see the following components in the [Virtualization sample app (aspnet/samples GitHub repository)](https://github.com/aspnet/samples/tree/master/samples/aspnetcore/blazor/Virtualization):
+
+* `Virtualize` component ([Shared/Virtualize.razor](https://github.com/aspnet/samples/blob/master/samples/aspnetcore/blazor/Virtualization/Shared/Virtualize.cs)) &ndash; A component written in C# that implements <xref:Microsoft.AspNetCore.Components.ComponentBase> to render a set of weather data rows based on user scrolling.
```

Comment:
Per the style guide, "Don’t use spaces on either side of an en dash."

```suggestion
* `Virtualize` component ([Shared/Virtualize.razor](https://github.com/aspnet/samples/blob/master/samples/aspnetcore/blazor/Virtualization/Shared/Virtualize.cs))&ndash;A component written in C# that implements <xref:Microsoft.AspNetCore.Components.ComponentBase> to render a set of weather data rows based on user scrolling.
```

Reply:
That's if the en-dash is used as a connector. Here, it's used as a layout element. Do you prefer a colon-space construct?

Reply:
I'll change to the colon-space construct.